src/libstd/path.rs

```diff
@@ -0,0 +1,2415 @@
+
+use core::prelude::*;
+
+use borrow::BorrowFrom;
+use cmp;
+use iter;
+use mem;
+use ops::{self, Deref};
+use string::CowString;
+use string::String;
+use vec::Vec;
+
+use ffi::{OsStr, OsString, AsOsStr};
+
+use self::platform::{is_sep, is_verbatim_sep, MAIN_SEP, MAIN_SEP_STR, parse_prefix, Prefix};
+
+#[cfg(unix)]
+#[path = "sys/unix/mod.rs"] mod sys;
+#[cfg(windows)]
+#[path = "sys/windows/mod.rs"] mod sys;
+
+////////////////////////////////////////////////////////////////////////////////
+// GENERAL NOTES
+////////////////////////////////////////////////////////////////////////////////
+//
+// Parsing in this module is done by directly transmuting OsStr to [u8] slices,
+// taking advantage of the fact that OsStr always encodes ASCII characters
+// as-is.  Eventually, this transmutation should be replaced by direct uses of
+// OsStr APIs for parsing, but it will take a while for those to become
+// available.
+
+////////////////////////////////////////////////////////////////////////////////
+// Platform-specific definitions
+////////////////////////////////////////////////////////////////////////////////
+
+// The following modules give the most basic tools for parsing paths on various
+// platforms. The bulk of the code is devoted to parsing prefixes on Windows.
+
+#[cfg(unix)]
+mod platform {
+    use core::prelude::*;
+    use ffi::OsStr;
+
+    #[inline]
+    pub fn is_sep(b: u8) -> bool {
+        b == b'/'
+    }
+
+    #[inline]
+    pub fn is_verbatim_sep(b: u8) -> bool {
+        b == b'/'
+    }
+
+    pub fn parse_prefix(_: &OsStr) -> Option<Prefix> {
+        None
+    }
+
+    #[derive(Copy, Clone, PartialEq)]
+    pub struct Prefix<'a>;
+
+    impl<'a> Prefix<'a> {
+        #[inline]
+        pub fn len(&self) -> usize { 0 }
+        #[inline]
+        pub fn is_verbatim(&self) -> bool { false }
+        #[inline]
+        pub fn is_drive(&self) -> bool { false }
+        #[inline]
+        pub fn has_implicit_root(&self) -> bool { false }
+    }
+
+    pub const MAIN_SEP: u8 = b'/';
+    pub const MAIN_SEP_STR: &'static str = "/";
+}
+
+#[cfg(windows)]
+mod platform {
+    use core::prelude::*;
+
+    use super::{Path, os_str_as_u8_slice, u8_slice_as_os_str};
+    use ffi::OsStr;
+    use ascii::*;
+
+    #[inline]
+    pub fn is_sep(b: u8) -> bool {
+        b == b'/' || b == b'\\'
+    }
+
+    #[inline]
+    pub fn is_verbatim_sep(b: u8) -> bool {
+        b == b'\\'
+    }
+
+    pub fn parse_prefix<'a>(path: &'a OsStr) -> Option<Prefix> {
+        use self::Prefix::*;
+        unsafe {
+            // The unsafety here stems from converting between &OsStr and &[u8]
+            // and back. This is safe to do because (1) we only look at ASCII
+            // contents of the encoding and (2) new &OsStr values are produced
+            // only from ASCII-bounded slices of existing &OsStr values.
+            let mut path = os_str_as_u8_slice(path);
+
+            if path.starts_with(b"\\\\") {
```

Comment:
I have a feeling I'm gonna start noticing this very quickly, but I might broadly recommend making liberal use of raw strings for `\` characters as it may help ease situations like this.